Firestore/core/src/firebase/firestore/remote/connectivity_monitor.cc

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/connectivity_monitor.h"
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+void ConnectivityMonitor::SetInitialStatus(NetworkStatus new_status) {
+  HARD_ASSERT(!status_.has_value(),
+              "SetInitialStatus should only be called once");
+  status_ = new_status;
+}
+
+void ConnectivityMonitor::MaybeInvokeCallbacks(NetworkStatus new_status) {
+  // TODO(varconst): cancel the enqueued operation in destructor or use
+  // shared/weak_ptr.
+  worker_queue_->Enqueue([this, new_status] {
```

Comment:
I'm not sure whether it's a good idea for the base class to be queue-aware.

Reply:
Is the design that callback should be done for each change of status or just dealing with the most recent status? Say, if the network disconnected and then recovered quickly, so quick that the callback on disconnected is still in queue waiting to be executed. Then should the callback gets disconnected and connected two events or gets nothing?